src/coreComponents/constitutive/fluid/multifluid/CO2Brine/functions/CO2Solubility.hpp

```diff
@@ -229,112 +241,169 @@ CO2SolubilityUpdate::compute( real64 const & pressure,
+    for( integer ic = 0; ic < 2; ic++ )
+    {
+      invDeterminantDeriv[ic] = invDeterminant*invDeterminant*(co2Solubility*watSolubilityDeriv[ic] + watSolubility*co2SolubilityDeriv[ic]);
+    }
   }
   else
   {
-    real64 const oneMinusCompFracInv = 1.0 / (1.0 - compFraction[m_CO2Index]);
-    Y = compFraction[m_CO2Index] * oneMinusCompFracInv;
-    dY_dCompFrac[m_CO2Index] = oneMinusCompFracInv * oneMinusCompFracInv;
-    dY_dCompFrac[m_waterIndex] = 0.0;
+    invDeterminant = 1.0 / minForDivision;
   }
 
-  auto setZero = []( real64 & val ){ val = 0.0; };
-  LvArray::forValuesInSlice( phaseFraction.derivs, setZero );
-  LvArray::forValuesInSlice( phaseCompFraction.derivs, setZero );
-
-  if( Y < solubility )
+  real64 x_co2 = co2Solubility * (z_wat - watSolubility * z_co2) * invDeterminant;
+  real64 x_co2Deriv[4]{ 0.0, 0.0, 0.0, 0.0 };
+  if( minForDivision < x_co2 )
   {
-    // liquid phase only
-
-    // 1) Compute phase fractions
-
-    phaseFraction.value[m_phaseLiquidIndex] = 1.0;
-    phaseFraction.value[m_phaseGasIndex] = 0.0;
-
-    // 2) Compute phase component fractions
-
-    phaseCompFraction.value[m_phaseGasIndex][m_CO2Index] = 1.0;
-    phaseCompFraction.value[m_phaseGasIndex][m_waterIndex] = 0.0;
-    for( localIndex ic = 0; ic < 2; ++ic )
+    // Pressure and temperature derivatives
+    for( integer ic = 0; ic < 2; ic++ )
     {
-      phaseCompFraction.value[m_phaseLiquidIndex][ic] = compFraction[ic];
+      x_co2Deriv[ic] = co2SolubilityDeriv[ic] * (z_wat - watSolubility * z_co2) * invDeterminant
+                       - co2Solubility * watSolubilityDeriv[ic] * z_co2 * invDeterminant
+                       + co2Solubility * (z_wat - watSolubility * z_co2) * invDeterminantDeriv[ic];
+    }
+    // Composition derivatives
+    x_co2Deriv[2+m_CO2Index] = -co2Solubility * watSolubility * invDeterminant;
+    x_co2Deriv[2+m_waterIndex] = co2Solubility * invDeterminant;
+  }
+  else
+  {
+    x_co2 = 0.0;
+  }
 
-      for( localIndex jc = 0; jc < 2; ++jc )
-      {
-        phaseCompFraction.derivs[m_phaseLiquidIndex][ic][Deriv::dC+jc] = (ic == jc ) ? 1.0 : 0.0;
-        phaseCompFraction.derivs[m_phaseGasIndex][ic][Deriv::dC+jc] = 0.0;
-      }
+  real64 y_wat = watSolubility * (z_co2 - x_co2);
+  real64 y_watDeriv[4]{ 0.0, 0.0, 0.0, 0.0 };
+  if( minForDivision < y_wat )
+  {
+    // Pressure and temperature derivatives
+    for( integer ic = 0; ic < 2; ic++ )
+    {
+      y_watDeriv[ic] = watSolubilityDeriv[ic] * (z_co2 - x_co2) - watSolubility * x_co2Deriv[ic];
     }
+    // Composition derivatives
+    y_watDeriv[2+m_CO2Index] = watSolubility*(1.0 - x_co2Deriv[2+m_CO2Index]);
+    y_watDeriv[2+m_waterIndex] = -watSolubility * x_co2Deriv[2+m_waterIndex];
   }
   else
   {
-    // two-phase flow
+    y_wat = 0.0;
+  }
 
-    // 1) Compute phase fractions
+  // Liquid and vapour phase fractions
+  real64 const L = x_co2 + z_wat - y_wat;
+  real64 const V = y_wat + z_co2 - x_co2; // = 1 - L;
```

Comment:
```suggestion
  real64 const V = 1 - L;// y_wat + z_co2 - x_co2;
```
Not for sure but this should be less numerical round-off sensible, isn't it ?